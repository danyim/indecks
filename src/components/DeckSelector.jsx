import React from 'react'
import styles from '../styles/components/DeckSelector.styl'

const propTypes = {
  // decks: React.PropTypes.array.isRequired,
  // excludedDeckIds: React.PropTypes.array,
  // maxSelected: React.PropTypes.number.isRequired,
  // minSelected: React.PropTypes.number.isRequired,
  // handleOnSelected: React.PropTypes.func.isRequired
}

const defaultProps = {
  maxSelected: 1,
  minSelected: 1
}

/**
 * Should be a simple pop-up of all the decks available to choose from,
 * optionally excluding (or greying out) a particular deck from the list.
 *
 * Input: decks to choose from, deck to exclude from selection, number of decks
 *   to allow for selection via [min, max]
 * Output: deck(s) selected from the list
 */
const DeckSelector = () => {
  const { decks } = this.props

  return (
    <div>
      <p>Yo, motherfucker, here are your decks:</p>
      {decks.map(x =>
        <div className={`${styles['deck-list-item']}`}>
          deck Id: {x.deckId}
        </div>
      )}
    </div>
  )
}

DeckSelector.propTypes = propTypes
DeckSelector.defaultProps = defaultProps

export default DeckSelector
