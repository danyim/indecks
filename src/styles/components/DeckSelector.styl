@import '../_variables';

.deck-list
  padding: 0
  margin: 0
  list-style-type: none
  li
    margin: 0
    font-size: 1.5rem
    transition: 0.2s all
  :global
    li.selected
      font-weight: bold
      font-size: 2rem
  &:focus
    outline: 0

.help
  margin-top: 2rem
  color: #c0c0c0
  font-size: 1.5rem
  // font-weight: bold
