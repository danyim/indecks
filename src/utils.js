export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

