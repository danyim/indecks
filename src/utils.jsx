import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
