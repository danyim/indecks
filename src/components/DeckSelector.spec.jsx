import toJson from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import DeckSelector from './DeckSelector'

const defaultProps = {
  decks: [
    {
      id: 'ABCDEFG',
      title: 'Deck title',
      description: 'Deck description',
      cards: [],
    },
    {
      id: '123455',
      title: 'Deck title',
      description: 'Deck description',
      cards: [],
    },
    {
      id: '121451',
      title: 'Deck title',
      description: 'Deck description',
      cards: [],
    },
  ],
  push: () => {},
  handleOnSelected: () => {},
}

describe('DeckSelector', () => {
  it('should render self and subcomponents', () => {
    const wrapper = mount(DeckSelector, defaultProps)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should list all decks available', () => {
    const { wrapper } = setup(DeckSelector, defaultProps)

    const listItem = wrapper.find('li')
    expect(listItem.length).toBe(defaultProps.decks.length)
  })
})
