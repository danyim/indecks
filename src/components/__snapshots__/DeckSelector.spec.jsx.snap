// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckSelector should render self and subcomponents 1`] = `
<DeckSelector
  decks={
    Array [
      Object {
        "cards": Array [],
        "description": "Deck description",
        "id": "ABCDEFG",
        "title": "Deck title",
      },
      Object {
        "cards": Array [],
        "description": "Deck description",
        "id": "123455",
        "title": "Deck title",
      },
      Object {
        "cards": Array [],
        "description": "Deck description",
        "id": "121451",
        "title": "Deck title",
      },
    ]
  }
  handleOnSelected={[Function]}
  push={[Function]}
>
  <div>
    <h2
      className="header"
    >
      Deck Selector
    </h2>
    <p
      className="help"
    >
      Use the up and down arrows to select a deck. Press enter to navigate to the deck.
    </p>
    <ul
      className="deck-list"
      onKeyDown={[Function]}
      role="presentation"
      tabIndex={-1}
    >
      <li
        className="selected"
        key="ABCDEFG"
        value="ABCDEFG"
      >
        Deck title
         
        <small>
          0
           cards
        </small>
      </li>
      <li
        className=""
        key="123455"
        value="123455"
      >
        Deck title
         
        <small>
          0
           cards
        </small>
      </li>
      <li
        className=""
        key="121451"
        value="121451"
      >
        Deck title
         
        <small>
          0
           cards
        </small>
      </li>
    </ul>
  </div>
</DeckSelector>
`
